x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -275,221 +284,109 @@ public void checkCompatibility(MappedFieldType otherFT, List<String> conflicts)
-                }
-                if (fragment.codePointCount(0, fragment.length()) <= NGRAM_SIZE) {
-                    tokens.add(fragment);
-                } else {
-                    // Break fragment into multiple Ngrams
-                    TokenStream tokenizer = WILDCARD_ANALYZER.tokenStream(name(), fragment);
-                    CharTermAttribute termAtt = tokenizer.addAttribute(CharTermAttribute.class);
-                    String lastUnusedToken = null;
-                    try {
-                        tokenizer.reset();
-                        boolean takeThis = true;
-                        // minimise number of terms searched - eg for "12345" and 3grams we only need terms
-                        // `123` and `345` - no need to search for 234. We take every other ngram.
-                        while (tokenizer.incrementToken()) {
-                            String tokenValue = termAtt.toString();
-                            if (takeThis) {
-                                tokens.add(tokenValue);
-                            } else {
-                                lastUnusedToken = tokenValue;
-                            }
-                            // alternate
-                            takeThis = !takeThis;
-                        }
-                        if (lastUnusedToken != null) {
-                            // given `cake` and 3 grams the loop above would output only `cak` and we need to add trailing
-                            // `ake` to complete the logic.
-                            tokens.add(lastUnusedToken);
-                        }
-                        tokenizer.end();
-                        tokenizer.close();
-                    } catch (IOException ioe) {
-                        throw new ElasticsearchParseException("Error parsing wildcard query pattern fragment [" + fragment + "]");
-                    }
-                }
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates, RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException("[regexp] queries cannot be executed when '" +
+                        ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false.");
             }
-
-            if (patternStructure.isMatchAll()) {
-                return new MatchAllDocsQuery();
+            failIfNotIndexed();
+            RegExp regex =new RegExp(value, flags);
+            try {
+                // TODO always lower case the ngram index and value? Could help with 
+                // a) speed (less ngram variations to explore on disk) and 
+                // b) use less disk space
+                Automaton automaton = regex.toAutomaton(maxDeterminizedStates);
+                return automatonToQuery(value, automaton);
+            } catch (AutomatonTooComplexException e) {
```

Comment:
Maybe we should just throw `IllegalArgumentException` instead of catch and rethrow?

Reply:
OTOH, you can actually mention regex here. So maybe it is worth it.